### PR TITLE
Rework controllers to use jwt token

### PR DESCRIPTION
**Pull request by ChernykhVyacheslav**
Rework controllers to use jwt token instead of user id

https://trello.com/c/SCKxQhnL/149-rework-controllers-to-use-jwt-token

### CODE REVIEW

File: controller/src/main/java/com/softserve/borda/config/jwt/JwtConvertor.java [hunk 1]
```diff
@@ -2,16 +2,19 @@
...
 
 @Service
 public class JwtConvertor {
 
-    private final com.softserve.borda.config.jwt.JwtProvider jwtProvider;
+    private final JwtProvider jwtProvider;
 
     private final JWTUserRepository jwtUserRepository;
 
+    private static final int JWT_START = 7;
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/config/jwt/JwtConvertor.java]: Does api use it somewhere else?
If not then remove that property.

File: controller/src/main/java/com/softserve/borda/controllers/TicketController.java [hunk 2]
```diff
@@ -264,11 +267,11 @@
         }
     }
 
-    @PostMapping("/tickets/{ticketId}/addMember/{userId}")
+    @PostMapping("/tickets/{ticketId}/addMember")
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/TicketController.java]: Bad resource naming.
Please read https://restfulapi.net/resource-naming/ doc and change naming.

File: controller/src/main/java/com/softserve/borda/controllers/TicketController.java [hunk 3]
```diff
@@ -280,11 +283,11 @@
         }
     }
 
-    @DeleteMapping("/tickets/{ticketId}/deleteMember/{userId}")
+    @DeleteMapping("/tickets/{ticketId}/deleteMember")
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/TicketController.java]: Bad resource naming.
Please read https://restfulapi.net/resource-naming/ doc and change naming.

File: controller/src/main/java/com/softserve/borda/controllers/UserController.java [hunk 4]
```diff
@@ -30,7 +30,7 @@
 
     private final JwtConvertor jwtConvertor;
 
-    @GetMapping
+    @GetMapping(value = "getAll")
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/UserController.java]: Bad resource naming.
Please read https://restfulapi.net/resource-naming/ doc and change naming.

File: controller/src/main/java/com/softserve/borda/controllers/UserController.java [hunk 5]
```diff
@@ -42,24 +42,11 @@
...
-        } catch (CustomEntityNotFoundException e) {
-            log.severe(e.getMessage());
-            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
-        }
-    }
-
-    @GetMapping("{id}")
-    public ResponseEntity<UserSimpleDTO> getUserById(@PathVariable String id) {
+    @GetMapping
+    public ResponseEntity<UserSimpleDTO> getUserByToken(@RequestHeader String authorization) {
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/UserController.java]: Can you explain how it works and for what does it serve?

File: controller/src/main/java/com/softserve/borda/controllers/UserController.java [hunk 6]
```diff
@@ -108,19 +95,21 @@
         }
     }
 
-    @PostMapping("update/{id}")
-    public UserSimpleDTO updateUser(@RequestBody final UserUpdateDTO userDTO, @PathVariable String id) {
-        var user = jwtConvertor.getUserByJWT(id);
+    @PostMapping("update")
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/UserController.java]: Bad resource naming.
Please read https://restfulapi.net/resource-naming/ doc and change naming.

File: controller/src/main/java/com/softserve/borda/controllers/UserController.java [hunk 7]
```diff
@@ -131,12 +120,13 @@ public UserSimpleDTO updateUser(@RequestBody final UserUpdateDTO userDTO, @PathV
         }
     }
 
-    @GetMapping("{userId}/boardsByRole/{boardRoleId}")
-    public ResponseEntity<List<BoardFullDTO>> getBoardsByUserIdAndBoardRoleId(@PathVariable Long userId,
+    @GetMapping("/boardsByRole/{boardRoleId}")
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/UserController.java]: Bad resource naming.
Please read https://restfulapi.net/resource-naming/ doc and change naming.

File: controller/src/main/java/com/softserve/borda/controllers/UserController.java [hunk 8]
```diff
@@ -146,11 +136,11 @@ public UserSimpleDTO updateUser(@RequestBody final UserUpdateDTO userDTO, @PathV
         }
     }
 
-    @GetMapping("/{userId}/comments")
-    public ResponseEntity<List<CommentDTO>> getCommentsByTicketId(@PathVariable Long userId) {
+    @GetMapping("/comments")
+    public ResponseEntity<List<CommentDTO>> getCommentsByUser(@RequestHeader String authorization) {
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/UserController.java]: For what does it serve?
What comments should it return?

File: service/src/main/java/com/softserve/borda/services/UserService.java [hunk 9]
```diff
@@ -12,15 +12,15 @@
     User createOrUpdate(User user);
     void deleteUserById(Long id);
 
-    List<Board> getBoardsByUserId(Long id);
+    List<Board> getBoardsByUser(User user);
 
-    List<Board> getBoardsByUserIdAndBoardRoleId(Long userId, Long roleId);
+    List<Board> getBoardsByUserAndBoardRoleId(User user, Long roleId);
```
Comment (achehrynets) [inline service/src/main/java/com/softserve/borda/services/UserService.java]: I think the previous format was better.

File: service/src/main/java/com/softserve/borda/services/UserService.java [hunk 10]
```diff
@@ -12,15 +12,15 @@
     User createOrUpdate(User user);
     void deleteUserById(Long id);
 
-    List<Board> getBoardsByUserId(Long id);
+    List<Board> getBoardsByUser(User user);
```
Comment (achehrynets) [inline service/src/main/java/com/softserve/borda/services/UserService.java]: I think the previous format was better.

File: service/src/main/java/com/softserve/borda/services/impl/UserServiceImpl.java [hunk 11]
```diff
@@ -64,18 +64,14 @@ public void deleteUserById(Long id) {
     }
 
     @Override
-    public List<Board> getBoardsByUserId(Long id) {
-        User user = userRepository.findById(id).orElseThrow(
-                () -> new CustomEntityNotFoundException(User.class));
+    public List<Board> getBoardsByUser(User user) {
         return user.getUserBoardRelations().stream().map(
```
Comment (achehrynets) [inline service/src/main/java/com/softserve/borda/services/impl/UserServiceImpl.java]: Do not fetch entity-related entities using getters, use the service layer instead.

File: service/src/main/java/com/softserve/borda/services/impl/UserServiceImpl.java [hunk 12]
```diff
@@ -64,18 +64,14 @@ public void deleteUserById(Long id) {
...
                 UserBoardRelation::getBoard
         ).collect(Collectors.toList());
     }
 
     @Override
-    public List<Board> getBoardsByUserIdAndBoardRoleId(Long userId, Long roleId) {
-        User user = userRepository.findById(userId).orElseThrow(
-                () -> new CustomEntityNotFoundException(User.class));
+    public List<Board> getBoardsByUserAndBoardRoleId(User user, Long roleId) {
         return user.getUserBoardRelations().stream()
```
Comment (achehrynets) [inline service/src/main/java/com/softserve/borda/services/impl/UserServiceImpl.java]: Do not fetch entity-related entities using getters, use the service layer instead.

File: service/src/main/java/com/softserve/borda/services/impl/UserServiceImpl.java [hunk 13]
```diff
@@ -106,7 +102,7 @@ public User deleteCommentFromUser(Long userId, Comment comment) {
     }
 
     @Override
-    public List<Comment> getAllCommentsByUserId(Long userId) {
-        return getUserById(userId).getComments();
+    public List<Comment> getAllCommentsByUser(User user) {
+        return user.getComments();
```
Comment (achehrynets) [inline service/src/main/java/com/softserve/borda/services/impl/UserServiceImpl.java]: Do not fetch entity-related entities using getters, use the service layer instead.

File: domain/src/main/java/com/softserve/borda/entities/User.java [hunk 14]
```diff
@@ -39,8 +39,6 @@
     private String lastName;
 
     private boolean enabled = true;
-
-    @Column(name = "avatar")
```
Comment (achehrynets) [inline domain/src/main/java/com/softserve/borda/entities/User.java]: Why did you remove mapping?

File: service/src/main/java/com/softserve/borda/services/UserService.java (hunk 10 above)
Comment (achehrynets) [reply service/src/main/java/com/softserve/borda/services/UserService.java]: Why info `UserService` returns `Board` entity?

File: service/src/main/java/com/softserve/borda/services/UserService.java (hunk 9 above)
Comment (achehrynets) [reply service/src/main/java/com/softserve/borda/services/UserService.java]: Why info `UserService` returns `Board` entity?

**Review by achehrynets — CHANGES_REQUESTED** [review]
Requested changes, waiting for answers.

File: controller/src/main/java/com/softserve/borda/config/jwt/JwtConvertor.java (hunk 1 above)
Comment (ChernykhVyacheslav) [reply controller/src/main/java/com/softserve/borda/config/jwt/JwtConvertor.java]: It uses it in the same file:
return jwtUserRepository.findById(jwt.substring(JWT_START)).orElseThrow(
                () -> new CustomEntityNotFoundException(User.class)).getUser();

We pass jwt token from header, which always has "Bearer {token}", so we cut off the starting part, leaving only token. Property was created to avoid usage of 'Magic numbers'

File: domain/src/main/java/com/softserve/borda/entities/User.java (hunk 14 above)
Comment (ChernykhVyacheslav) [reply domain/src/main/java/com/softserve/borda/entities/User.java]: Not sure. Must have been accidental change, my bad. Brought it back

File: service/src/main/java/com/softserve/borda/services/impl/UserServiceImpl.java (hunk 13 above)
Comment (ChernykhVyacheslav) [reply service/src/main/java/com/softserve/borda/services/impl/UserServiceImpl.java]: Fixed in ReworkService branch/ticket

File: service/src/main/java/com/softserve/borda/services/impl/UserServiceImpl.java (hunk 12 above)
Comment (ChernykhVyacheslav) [reply service/src/main/java/com/softserve/borda/services/impl/UserServiceImpl.java]: Fixed in ReworkService branch/ticket

File: service/src/main/java/com/softserve/borda/services/impl/UserServiceImpl.java (hunk 11 above)
Comment (ChernykhVyacheslav) [reply service/src/main/java/com/softserve/borda/services/impl/UserServiceImpl.java]: Fixed in ReworkService branch/ticket

File: service/src/main/java/com/softserve/borda/services/UserService.java (hunk 10 above)
Comment (ChernykhVyacheslav) [reply service/src/main/java/com/softserve/borda/services/UserService.java]: Yes, returned to previous version. Fixed in ReworkService branch/ticket

File: controller/src/main/java/com/softserve/borda/controllers/UserController.java (hunk 6 above)
Comment (ChernykhVyacheslav) [reply controller/src/main/java/com/softserve/borda/controllers/UserController.java]: Fixed in FixResourcesPaths branch/ticket, already merged in main

File: controller/src/main/java/com/softserve/borda/controllers/UserController.java (hunk 7 above)
Comment (ChernykhVyacheslav) [reply controller/src/main/java/com/softserve/borda/controllers/UserController.java]: Fixed in FixResourcesPaths branch/ticket, already merged in main

File: controller/src/main/java/com/softserve/borda/controllers/UserController.java (hunk 5 above)
Comment (ChernykhVyacheslav) [reply controller/src/main/java/com/softserve/borda/controllers/UserController.java]: The idea is to get user by authorization token, rather than by id. So it would not be necessary to pass user id, but rather use JWT token already being passed. This way, we can be sure that the one who is requesting user info is the user himself.
Every request sent to our server (except authorization and registration ones) requires JWT token to be passed in a header in the following format:
"Authorization": "Bearer {token}". Then we pass the string "Bearer {token}" into JWTConverter, where token gets extracted and used to receive corresponding user from our database.

File: controller/src/main/java/com/softserve/borda/controllers/UserController.java (hunk 4 above)
Comment (ChernykhVyacheslav) [reply controller/src/main/java/com/softserve/borda/controllers/UserController.java]: Fixed in FixResourcesPaths branch/ticket, already merged in main

File: controller/src/main/java/com/softserve/borda/controllers/TicketController.java (hunk 3 above)
Comment (ChernykhVyacheslav) [reply controller/src/main/java/com/softserve/borda/controllers/TicketController.java]: Fixed in FixResourcesPaths branch/ticket, already merged in main

File: controller/src/main/java/com/softserve/borda/controllers/TicketController.java (hunk 2 above)
Comment (ChernykhVyacheslav) [reply controller/src/main/java/com/softserve/borda/controllers/TicketController.java]: Fixed in FixResourcesPaths branch/ticket, already merged in main

File: controller/src/main/java/com/softserve/borda/config/jwt/JwtConvertor.java (hunk 1 above)
Comment (ChernykhVyacheslav) [reply controller/src/main/java/com/softserve/borda/config/jwt/JwtConvertor.java]: Fixed

**Review by jules-n — APPROVED** [review]
Looks right

**Review by HorbatenkoYehor — APPROVED** [review]
+1
